### PR TITLE
MOSIP-28139 added handleCtkTokenFlow method

### PR DESCRIPTION
**Pull request by mayuradesh**

### CODE REVIEW

**Review by mahammedtaheer — CHANGES_REQUESTED** [review]

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 1]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+	/**
+	 * This is custom fail-safe handling added only for Compliance Toolkit, to
+	 * enable ABIS data share testing.
+	 */
+	private void handleCtkTokenFlow(HttpServletRequest httpServletRequest, String token) {
+		String ctkTestCaseId = null;
+		String ctkTestRunId = null;
+		Map<String, String[]> requestParams = httpServletRequest.getParameterMap();
+
+		for (Map.Entry<String, String[]> entry : requestParams.entrySet()) {
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:299]: Updated code

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 2]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+				ctkTestCaseId = entry.getValue()[0];
+				LOGGER.debug("Recvd ctkTestCaseId {}", ctkTestCaseId);
+			}
+			if (AuthAdapterConstant.CTK_TEST_RUN_ID.equals(entry.getKey()) && entry.getValue().length > 0) {
+				ctkTestRunId = entry.getValue()[0];
+				LOGGER.debug("Recvd ctkTestRunId {}", ctkTestRunId);
+			}
+		}
+		if (ctkTestCaseId != null && ctkTestRunId != null) {
+			String ctkSaveUrl = environment.getProperty("mosip.compliance.toolkit.saveDataShareToken.url");
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:310]: done

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 3]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+			// add the token first
+			HttpHeaders headers = new HttpHeaders();
+			headers.add(AuthAdapterConstant.AUTH_HEADER_COOKIE, AuthAdapterConstant.AUTH_HEADER + token);
+			headers.setContentType(MediaType.APPLICATION_JSON);
+			// create request
+			Map<String, String> valueMap = new HashMap<String, String>();
+			valueMap.put(AuthAdapterConstant.PARTNER_ID, partnerId);
+			valueMap.put(AuthAdapterConstant.CTK_TEST_CASE_ID, ctkTestCaseId);
+			valueMap.put(AuthAdapterConstant.CTK_TEST_RUN_ID, ctkTestRunId);
+			valueMap.put(AuthAdapterConstant.TOKEN, token);
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:340]: Updated code

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 4]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+			RequestWrapper<Object> requestWrapper = new RequestWrapper<>();
+			requestWrapper.setId("mosip.toolkit.abis.datashare.token");
+			requestWrapper.setVersion("1.0");
+			requestWrapper.setRequesttime(LocalDateTime.now());
+			requestWrapper.setRequest(valueMap);
+			LOGGER.debug("Calling Compliance Toolkit requestWrapper: " + requestWrapper);
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:346]: updated code

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 5]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+			// ResponseWrapper<String> response = null;
+			ResponseEntity<ResponseWrapper<String>> responseEntity = null;
+			try {
+				HttpEntity<RequestWrapper<Object>> requestEntity = new HttpEntity<>(requestWrapper, headers);
+				String tokenUrl = new StringBuilder(ctkSaveUrl).toString();
+				String ctkInvalidateTestCaseId = environment.getProperty("mosip.compliance.toolkit.invalidateDataShareToken.testCaseId");
+				if (ctkInvalidateTestCaseId != null && ctkInvalidateTestCaseId.equals(ctkTestCaseId)) {
+					tokenUrl = new StringBuilder(ctkInvalidateUrl).toString();
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:354]: added @value annotation
Two separate urls perform two separate logic
in one , token is expired before saving to db and in one token is directly save in db

File: kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java [hunk 6]
```diff
@@ -263,5 +286,84 @@ private String getApplicationName(Environment environment) {
...
+				}
+				LOGGER.debug("Calling Compliance Toolkit URL: " + tokenUrl);
+				responseEntity = restTemplate.exchange(tokenUrl, HttpMethod.POST, requestEntity,
+						new ParameterizedTypeReference<ResponseWrapper<String>>() {
+						});
+				ResponseWrapper<String> body = responseEntity.getBody();
+				LOGGER.debug("Response from Compliance Toolkit: " + body.getResponse());
+			} catch (Exception e) {
+				// This is FailSafe, so just log the err
+				LOGGER.error("error connecting to compliance toolkit {}", e.getLocalizedMessage());
```
[the remark this answers is not in the document]
Comment (mayuradesh) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java:364]: done

[the remark this answers is not in the document]
Comment (mahammedtaheer) [reply kernel/kernel-auth-adapter/src/main/java/io/mosip/kernel/auth/defaultadapter/filter/AuthFilter.java]: getStackTrace() method returns StackTraceElement object array, the log statement convert it to string which will not give stacktrack. Please use "error(String message, Throwable throwable)" method to print stacktrack in logs.

**Review by mahammedtaheer — APPROVED** [review]

**Pull request merged**